### ISSUE TITLE
malfeasance: add machine fingerprint when post data/smesher key is generated

### ISSUE THREAD

**Issue by dshulyak** [spacemeshos/go-spacemesh#4856]
we want to prevent accidental private key reuse across several machines on copying post data. 
if there is a fingerprint from unique non-ephemeral data we can enforce that it remains the same across go-spacemesh restarts.
if it is not the same we will ask user to acknowledge that the state was indeed moved and either delete specified fingerprint (or app can help with it). this fingerprint should also enable users to run docker without additional issues.

fingerprint should contain unique fields specific to hardware:
- motherboard serial number (/sys/class/dmi/id/board_serial)
- macaddress (ip a)
- ? 

in particular on linux there is already a machine id, i assume generated from such sources as above and more. 

on startup node should compute fingerprint from hardware and read fingerprint file from disk, if it doesn't match the FATAL error should be returned to user with instructions on how to bypass it. several options:

- specify a file that has to be removed 
- provide link in the error message to instructions
- request to add option to config that will recompute file, this is not so good as people likely will add it and forger

besides that we need to make sure that:
- initialization on multiple machines is working
- node needs to have a defined behavior when files are not found: regenerate / create new ones.
- a resulting code MUST work in containerized setups.

fingerprint should be generated together with key, specifically as an upgrade path we should always generate it if not found, and after couple of weeks start failing if fingerprint is not found, but key exists.

**Comment by dshulyak** [comment]
@pigmej shared a concern that if you generated data and key with postcli, then you may copy it from that place to multiple machines. we want to fail if there is no fingerprint, but for 1 week it will be slightly more risky to do this kind of stuff. beside that people that are using postcli should be mindful of doing such things anyway

**Comment by lrettig** [comment]
Regarding how we calculate the fingerprint, I propose the following:

- on systems/operating systems that already have something akin to a "serial number", use this - but in most cases this seems to require superuser/admin access (e.g., [`dmidecode -s system-uuid` on Linux](https://serverfault.com/questions/733069/i-need-a-command-to-create-a-unique-fingerprint-from-unix-machine)) so I'm not sure how helpful this is; otherwise:
- use the UUID (or its equivalent on other operating systems) of the filesystem where the PoST data resides - this is also probably sufficient on its own; otherwise:
- use the OS/platform (e.g., `linux-amd64`), and
- the filesystem location of the PoST data, and
- the MAC address of the default interface and/or the UUID of the root filesystem (which doesn't require root access)

[How to fingerprint a Unix System [closed]](https://unix.stackexchange.com/questions/323371/how-to-fingerprint-a-unix-system) lists a few other options but I don't think we need to go that far.

One more thing to consider: we need to make sure that this works in a dockerized setup too, so that, e.g., running the same node in two docker containers simultaneously on the same system would produce an error. I'm not sure what "serial number" or UUIDs would appear in a dockerized setup, we need to check this.

**Comment by lrettig** [comment]
> @pigmej shared a concern that if you generated data and key with postcli, then you may copy it from that place to multiple machines. we want to fail if there is no fingerprint, but for 1 week it will be slightly more risky to do this kind of stuff. beside that people that are using postcli should be mindful of doing such things anyway

I don't understand - are you suggesting that `postcli` shouldn't add a fingerprint? Shouldn't we always add a fingerprint regardless of how/where the PoST data was generated?

**Comment by dshulyak** [comment]
> are you suggesting that postcli shouldn't add a fingerprint? Shouldn't we always add a fingerprint regardless of how/where the PoST data was generated?

if user is forced to generate fingerprint when node starts then it doesn't matter if postcli generated fingerprint or not. i don't mind if someone will choose to generate fingerprint in postcli too and ensure that fingerprint is consistent with go-spacemesh. note that they will probably have to regenerate it because of using postcli on different machine, but again this doesn't matter, as long as node forces them to do it

**Comment by lrettig** [comment]
> on startup node should compute fingerprint from hardware and read fingerprint file from disk, if it doesn't match the FATAL error should be returned to user with instructions on how to bypass it. several options:
> 
> * specify a file that has to be removed
> * provide link in the error message to instructions
> * request to add option to config that will recompute file, this is not so good as people likely will add it and forger

I'd favor:

- display a big, all caps error message that warns about the situation, something along the lines of, "use caution before proceeding, make sure you know what you're doing, or your identity may be disqualified and you'll miss rewards"
- the message should contain a link to a document (e.g., wiki page) with more information and instructions
- in go-spacemesh, require the user to add a CLI flag with a specific layer number and restart. the node would then do two things differently: 1. ignore self-signed messages up to that layer (which must be a historic layer), and 2. regenerate the fingerprint. it would look like this: `--unsafe-resync-to <layerid>`

**Comment by ndcc5** [comment]
I think this will cause more problems through hardware inspection, 1. Almost all the generated data (GPU PC) will not be on the same machine as the running node (CPU PC).2. Many people may transfer data to other PCS when scanning disks. For faster scanning disks.

My suggestion is: When running the node, the node ID is broadcast throughout the network, and every time the node logs in, it is found that there are duplicate ids in the network.

**Comment by lrettig** [comment]
> When running the node, the node ID is broadcast throughout the network, and every time the node logs in, it is found that there are duplicate ids in the network.

We discussed this possibility. This is more complex than it sounds. It requires creating a new type of gossip message, and gossip messages can be abused (e.g., flooding). So we're creating new problems - how do we prevent spam, and punish nodes that spam these messages? Another problem is that nodes don't "log in" to anything in a decentralized network. They just begin signing and broadcasting messages, and if two nodes with identical IDs do this at the same time, that's equivocation.

> 1. Almost all the generated data (GPU PC) will not be on the same machine as the running node (CPU PC).2. Many people may transfer data to other PCS when scanning disks.

There's no issue with either case. In both cases, the user would get a warning when they first tried to smesh on the new system. They'd read the message, make sure they had shut down the old node (which is the whole point), then override the warning.

Thanks for the feedback!

**Comment by hengaoxing233** [comment]
What if my computer is damaged and I have to delete POST? Can't I move to another computer?

**Comment by lrettig** [comment]
> What if my computer is damaged and I have to delete POST? Can't I move to another computer?

Yes. As I wrote above, you'll get a warning the first time you run Smapp/go-spacemesh on the new computer. If you've already shut down the node/deleted the files on the old computer, you can override the warning.

**Comment by lrettig** [comment]
lowering priority as equivocation has been much less of an issue the past couple of months since we stepped up communication about this.
